python/pipeline/utils/clocktools.py

```diff
@@ -0,0 +1,996 @@
+                                             at indices instead of target indices
+                                             NOTE: NaNs in source signal are linearly interpolated over and then
+                                                   filled back in after interpolation onto target indices.
+                                     - False: Return corresponding indices for target
+
+                drop_single_idx: Boolean with the following behavior
+                                     - True: Drop any signal fragments which would create a signal of length 1.
+                                     - False: Raise an error and stop if any list of indices leads to a signal
+                                              of length 1.
+                                    ex. Source IDX [1,2,...,300] on a 500HZ recording will only correspond to
+                                        target IDX [1] if target is recorded at 1Hz.
+
+                debug: Set function to print helpful debug text while running
+
+
+        Returns:
+
+                requested_array: Numpy array of corresponding indices in target type or interpolated source signal 
+                                 during target indices.
+                                 
+        Warnings:
+        
+                * NaN refilling for source signal will only refill values if NaNs stretch for multiple indices on 
+                  target clock
+                  
+                * Recording points where the time value is NaN are dropped from analysis/processing
+    """
+
+    ##
+    ## Fetch source and target times, along with converting between Stimulus or Behavior clock if needed
+    ##
+
+    source_times, target_times = fetch_timing_data(
+        scan_key, source_type, target_type, debug
+    )
+
+    ##
+    ## Convert indices to a list of numbers if argument equals None or a Boolean mask
+    ##
+
+    if indices is None:
+        indices = np.arange(len(source_times))
+    elif type(indices[0]) == bool:
+        indices = np.where(indices)[0]
+    else:
+        ## Check for duplicates if manually entered
+        if len(np.unique(indices)) != len(indices):
+            msg = (
+                f"Duplicate entries found for provided indice array! "
+                f"Try to fix the error or use np.unique() on indices array."
+            )
+            raise PipelineException(msg)
+
+    ##
+    ## Convert source indices to time boundaries, then convert time boundaries into target indices
+    ##
+
+    ## Convert indices into start/end times for each continuous fragment (incrementing by 1)
+    time_boundaries = find_time_boundaries(indices, source_times, drop_single_idx)
+    target_indices = []
+    single_idx_count = 0
+
+    ## Loop through start & end times and create list of indices corresponding to that block of time
+    for [start, end] in time_boundaries:
```

Comment:
would be really nice if these next ~20 lines of code could be put into a function, that way you could reuse them all across the convert_clocks_*_to_*